src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+    if (b->requested > 0)
+    {
+        number_of_samples = SPA_MIN(number_of_samples, b->requested * state->stride); // because this value is given in frames, and we want samples
+    }
+#endif
+
+    // if there's something to be read from the ring buffer at this time, we do so now
+    if (available_samples > 0)
+    {
+        available_for_loading = SPA_MIN(available_samples, number_of_samples);
+    }
+    else
+    {
+        available_for_loading = 0;
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = number_of_samples - available_for_loading;
+    if (available_for_loading > 0)
+    {
+        spa_ringbuffer_read_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, read_index, destination_memory, available_for_loading);
+        spa_ringbuffer_read_update(&state->rb, read_index + available_for_loading);
+    }
+// if the required pipewire version doesn't match for requested to be available, we may have had plenty more time to buffer some more data, and instead we fill the whole buffer, while reading everything from the ringbuffer prematurely
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        if (must_load_with_silence > 0)
+        {
+            memset(destination_memory + available_for_loading, 0, must_load_with_silence);
+        }
+    }
+#endif
+    // now, we should have most of the data ready, we fill in some metadata and push the buffer object to pipewire
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->size = number_of_samples;
+    buf->datas[0].chunk->stride = state->stride;
+    pw_stream_queue_buffer(state->stream, b);
+    // signal to the main thread that data can be written again
+    spa_system_eventfd_write(pw_thread_loop_get_loop(state->loop)->system, state->eventfd_number, 42);
+}
+// pipewire internal: structure describing what kind of events we subscribe to
+// For now, this is only on_process
+static const struct pw_stream_events stream_events = {
+    PW_VERSION_STREAM_EVENTS,
+    .process = on_process,
+};
+
+//  open a pipewire connection and initialize a non-blocking main loop
+//  pars[1] contains the name of the sink
+static AudioID *pipewire_open(void **pars)
+{
+    module_state *state = (module_state *)malloc(sizeof(module_state));
+    pw_init(0, NULL);
+    state->loop = pw_thread_loop_new("pipewire audio thread", NULL);
+    state->rb = SPA_RINGBUFFER_INIT();
+    state->sample_buffer = malloc(SAMPLE_BUFFER_SIZE);
+    // we don't know if the params array belongs to us
+    // it  could be freed by speech dispatcher at a later date, which means we could be storing a pointer to freed memory
+    // to ensure that the params array belongs to us and only we can free it, we will duplicate the string contained in params[1], even if such an operation would have not been necesary in the end
+    state->sink_name = strdup(pars[1]);
```

Comment:
`pw_stream_new_simple` makes a copy of the `name` being passed, so no need to strdup it.